ec/ecresource/deploymentresource/deployment/v2/deployment.go

```diff
@@ -0,0 +1,566 @@
+	Version               string                                   `tfsdk:"version"`
+	Region                string                                   `tfsdk:"region"`
+	DeploymentTemplateId  string                                   `tfsdk:"deployment_template_id"`
+	Name                  string                                   `tfsdk:"name"`
+	RequestId             string                                   `tfsdk:"request_id"`
+	ElasticsearchUsername string                                   `tfsdk:"elasticsearch_username"`
+	ElasticsearchPassword string                                   `tfsdk:"elasticsearch_password"`
+	ApmSecretToken        *string                                  `tfsdk:"apm_secret_token"`
+	TrafficFilter         []string                                 `tfsdk:"traffic_filter"`
+	Tags                  map[string]string                        `tfsdk:"tags"`
+	Elasticsearch         *elasticsearchv2.Elasticsearch           `tfsdk:"elasticsearch"`
+	Kibana                *kibanav2.Kibana                         `tfsdk:"kibana"`
+	Apm                   *apmv2.Apm                               `tfsdk:"apm"`
+	IntegrationsServer    *integrationsserverv2.IntegrationsServer `tfsdk:"integrations_server"`
+	EnterpriseSearch      *enterprisesearchv2.EnterpriseSearch     `tfsdk:"enterprise_search"`
+	Observability         *observabilityv2.Observability           `tfsdk:"observability"`
+}
+
+// Nullify Elasticsearch topologies that have zero size and are not specified in plan
+func (dep *Deployment) NullifyNotUsedEsTopologies(ctx context.Context, esPlan *elasticsearchv2.ElasticsearchTF) {
+	if dep.Elasticsearch == nil {
+		return
+	}
+
+	if esPlan == nil {
+		return
+	}
+
+	dep.Elasticsearch.HotTier = nullifyUnspecifiedZeroSizedTier(esPlan.HotContentTier, dep.Elasticsearch.HotTier)
+
+	dep.Elasticsearch.WarmTier = nullifyUnspecifiedZeroSizedTier(esPlan.WarmTier, dep.Elasticsearch.WarmTier)
+
+	dep.Elasticsearch.ColdTier = nullifyUnspecifiedZeroSizedTier(esPlan.ColdTier, dep.Elasticsearch.ColdTier)
+
+	dep.Elasticsearch.FrozenTier = nullifyUnspecifiedZeroSizedTier(esPlan.FrozenTier, dep.Elasticsearch.FrozenTier)
+
+	dep.Elasticsearch.MlTier = nullifyUnspecifiedZeroSizedTier(esPlan.MlTier, dep.Elasticsearch.MlTier)
+
+	dep.Elasticsearch.MasterTier = nullifyUnspecifiedZeroSizedTier(esPlan.MasterTier, dep.Elasticsearch.MasterTier)
+
+	dep.Elasticsearch.CoordinatingTier = nullifyUnspecifiedZeroSizedTier(esPlan.CoordinatingTier, dep.Elasticsearch.CoordinatingTier)
+}
+
+func nullifyUnspecifiedZeroSizedTier(tierPlan types.Object, tier *elasticsearchv2.ElasticsearchTopology) *elasticsearchv2.ElasticsearchTopology {
+
+	if tierPlan.IsNull() && tier != nil {
+
+		size, err := converters.ParseTopologySize(tier.Size, tier.SizeResource)
+
+		// we can ignore returning an error here - it's handled in readers
+		if err == nil && size != nil && size.Value != nil && *size.Value == 0 {
```

Comment:
Are we reintroducing https://github.com/elastic/terraform-provider-ec/issues/471 here? Should we also be checking if the tier is autoscaleable?

Reply:
I don't think so. The current approach is to specify all tiers (even if they are empty) if they have non-zero size in the used template or can be scaled up (that is the case when autoscaling is enabled). So in case of #471, the TF configuration can look like 

```
resource "ec_deployment" "test-471" {
  name                   = "test_471"
  region                 = var.ec_region
  version                = data.ec_stack.latest.version
  deployment_template_id = "aws-storage-optimized-v2"

  elasticsearch = {
    autoscale = "true"

    hot = {
      size       = "1g"
      zone_count = 3

      autoscaling = {
        max_size = "8g"
      }
    }

    ml = {
      zone_count = 1

      autoscaling = {
        min_size = "0g"
        max_size = "60g"
      }
    }

    warm = {
      autoscaling = {}
    }

    cold = {
      autoscaling = {}
    }

    frozen = {
      autoscaling = {}
    }
    
  }

  kibana = {
    size       = "1g"
    zone_count = 2
  }

  lifecycle {
    ignore_changes = [
      elasticsearch.hot.size,
      elasticsearch.warm.size,
      elasticsearch.cold.size,
      elasticsearch.frozen.size,
      elasticsearch.ml.size,
    ]
  }  

}

```

[README](https://github.com/dimuon/terraform-provider-ec/blob/feature/530/migrate-to-plugin-framework/README.md#moving-to-tf-framework-and-schema-change-for-ec_deployment-resource) and [docs](https://github.com/dimuon/terraform-provider-ec/blob/feature/530/migrate-to-plugin-framework/docs/resources/ec_deployment.md#tiered-deployment-with-autoscaling-enabled) mentions this approach. Still any suggestions are welcomed.